### ISSUE TITLE
TF2.0, how to include feature processing code in saved model.

### ISSUE THREAD

**Issue by shangh1** [tensorflow/tensorflow#43373]
`<em>Please` make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes.
- OS Platform and Distribution : macOS Catalina 10.15.3
- TensorFlow installed from : binary
- TensorFlow version : 2.2.0
- Python version: 3.7.3

Hi, I have a working code (shown below): it processed the raw input (in tensor functions); then the processed input is used as the input of the keras model. I'm able to save the model, and also can make inference after loading model. But my problem is: this feature processing is done before inputing to the keras model, so it is not part of the saved model. When I run inference, I need to process the raw input first, then run model.predict on the processed data.

Because we would like to put the model on production, we need to include the feature processing code into the saved model, so the raw data can be fed into the saved model directly. I've tried different ways to make it work, but all failed. Please help, thanks!

Note that in the following code, 'FeatureProcess' is the function processing the raw input, I don't show the details of this 'FeatureProcess' function due to privacy reason. Inside this 'FeatureProcess' function is all tensor functions and manipulations.

def line_to_multiple_features(inputString, mode):
  inputString = [inputString]
  batch_ids, input_ids_long, labels, input_ids = FeatureProcess(inputString)
  d = {'ids':batch_ids[0], 'input_ids_long':input_ids_long[0], 'input_ids':input_ids[0]}
  if mode == 'train':
    return d, labels[0]
  else:
    return d

def get_unbatched_feature_label(filename, mode, epochs, feature_delimiter):
  dataset = tf.data.TextLineDataset(filename)
  if mode == 'train':
    dataset = dataset.repeat(epochs).shuffle(10000)
    # convert every line to features
  unbatch_parse_data = dataset.map(lambda x: line_to_multiple_features(x))
  return unbatch_parse_data

# Dataset for input data
def get_feature_label(_input):
  input_data = ctx.absolute_path(_input)
  input_files = tf.data.Dataset.list_files(input_data)
  datasets_unbatched = get_unbatched_feature_label(input_files, args.mode, args.epochs, '\t')
  data_feature_label = datasets_unbatched.batch(GLOBAL_BATCH_SIZE)
  return data_feature_label

def build_keras_model():
  vocab_size = 100000 
  embedding_dim = args.hidden_size
  num_filters = 512
  filter_sizes = [2,3,4]
  drop = 0.5

  inputs = {'ids':Input(shape=(4, ), dtype='string'),
              'input_ids_long':Input(shape=(args.seq_length_long, ), dtype='int32'),
              'input_ids':Input(shape=(args.seq_length, ), dtype='int32'))
             }

  embedding_short = Embedding(input_dim=vocab_size, output_dim=embedding_dim, 
  input_length=args.seq_length)(inputs['input_ids'])
  embedding_long = Embedding(input_dim=vocab_size, output_dim=embedding_dim, 
  input_length=args.seq_length_long)(inputs['input_ids_long'])
  reshape_short = Reshape((args.seq_length, embedding_dim, 1))(embedding_short)
  reshape_long = Reshape((args.seq_length_long, embedding_dim, 1))(embedding_long)
  conv_0_short = Conv2D(num_filters, kernel_size=(filter_sizes[0], embedding_dim), padding='valid', kernel_initializer='normal', activation='relu')(reshape_short)
  conv_1_short = Conv2D(num_filters, kernel_size=(filter_sizes[1], embedding_dim), padding='valid', kernel_initializer='normal', activation='relu')(reshape_short)
  conv_2_short = Conv2D(num_filters, kernel_size=(filter_sizes[2], embedding_dim), padding='valid', kernel_initializer='normal', activation='relu')(reshape_short)

  conv_0_long = Conv2D(num_filters, kernel_size=(filter_sizes[0], embedding_dim), padding='valid', kernel_initializer='normal', activation='relu')(reshape_long)
  conv_1_long = Conv2D(num_filters, kernel_size=(filter_sizes[1], embedding_dim), padding='valid', kernel_initializer='normal', activation='relu')(reshape_long)
  conv_2_long = Conv2D(num_filters, kernel_size=(filter_sizes[2], embedding_dim), padding='valid', kernel_initializer='normal', activation='relu')(reshape_long)

  maxpool_0_short = MaxPool2D(pool_size=(args.seq_length - filter_sizes[0] + 1, 1), strides=(1,1), padding='valid')(conv_0_short)
  maxpool_1_short = MaxPool2D(pool_size=(args.seq_length - filter_sizes[1] + 1, 1), strides=(1,1), padding='valid')(conv_1_short)
  maxpool_2_short = MaxPool2D(pool_size=(args.seq_length - filter_sizes[2] + 1, 1), strides=(1,1), padding='valid')(conv_2_short)

  maxpool_0_long = MaxPool2D(pool_size=(args.seq_length_long - filter_sizes[0] + 1, 1), strides=(1,1), padding='valid')(conv_0_long)
  maxpool_1_long = MaxPool2D(pool_size=(args.seq_length_long - filter_sizes[1] + 1, 1), strides=(1,1), padding='valid')(conv_1_long)
  maxpool_2_long = MaxPool2D(pool_size=(args.seq_length_long - filter_sizes[2] + 1, 1), strides=(1,1), padding='valid')(conv_2_long)

  concatenated_tensor = Concatenate(axis=1)([maxpool_0_short, maxpool_1_short, maxpool_2_short, maxpool_0_long, maxpool_1_long, maxpool_2_long])
  flatten = Flatten()(concatenated_tensor)
  dropout = Dropout(drop)(flatten)
  logits = Dense(units=args.num_classes)(dropout)
  pred_probs = tf.keras.activations.sigmoid(logits)
  model = Model(inputs=inputs, outputs=pred_probs)

  def get_loss_fn():
    def get_loss(labels, outputs):
      logits = tf.math.log((outputs+0.0000001) / ( 1 - outputs+0.0000001))
      loss = tf.nn.sigmoid_cross_entropy_with_logits(logits=logits, labels=labels)
      cost = tf.reduce_mean(
                tf.reduce_sum(loss, axis=1)
            )
      return cost
    return get_loss

  adam = Adam(lr=0.001)
  model.compile(optimizer=adam,
                  loss=get_loss_fn(),
                  metrics=['accuracy'])
  return model

tf.io.gfile.makedirs(args.model_dir)
filepath = args.model_dir # + "/weights-{epoch:04d}"
callbacks = [tf.keras.callbacks.ModelCheckpoint(filepath=filepath, monitor="val_accuracy", verbose=1, save_best_only=True, mode='auto', save_freq=10)]
steps_per_epoch = 60000 / GLOBAL_BATCH_SIZE * 0.9

if args.mode == 'train':
  with strategy.scope():
    multi_worker_model = build_keras_model()
  train_feature_label = get_feature_label(args.input)
  validation_feature_label = get_feature_label(args.val_path)
  multi_worker_model.fit(x=train_feature_label, validation_data=validation_feature_label, 
                                        epochs=args.epochs, steps_per_epoch=steps_per_epoch, callbacks=callbacks)
  results = multi_worker_model.evaluate(validation_feature_label, verbose=2)
  multi_worker_model.save(args.model_dir)
else: 
  print('now is inference mode')
  test_feature = get_feature_label(args.test_path)
  reconstructed_model = tf.keras.models.load_model(args.model_dir, compile=False)
  for elem in test_feature:
    y_pred_elem = reconstructed_model.predict(elem)

**Comment by shangh1** [comment]
sorry I put indented spaces for the code, but the indents were all gone after I submitted it.

**Comment by DachuanZhao** [comment]
> sorry I put indented spaces for the code, but the indents were all gone after I submitted it.

I have the same [problem](https://github.com/tensorflow/tensorflow/issues/42552) . They suggest me use ```tf.keras.layers.experimental.preprocessing```.  : )

**Comment by Saduf2019** [comment]
@shangh1 
Please share a colab gist with error reported.

**Comment by shangh1** [comment]
@DachuanZhao @Saduf2019  Thank you both! I'll try tf.keras.layers.experimental.preprocessing, and update you.

**Comment by shangh1** [comment]
Hi @Saduf2019 @DachuanZhao, my issue is similar as this [problem](https://github.com/tensorflow/tensorflow/issues/42552) brought up by @DachuanZhao. But I just looked at functions in tf.keras.layers.experimental.preprocessing, it only has around 20 classes, seems very limited. My data processing code is written all in tf-functions, and I'm not able to convert all tf-functions in my data processing code to tf.keras.layers.experimental.preprocessing. Some of my tf functions in data processing code include: tf.strings.split, tf.lookup.StaticHashTable, tf.strings.to_number, tf.sparse.SparseTensor, tf.compat.v1.sparse_to_dense, tf.fill, tf.concat, tf.not_equal, tf.cast.

**Comment by shangh1** [comment]
@Saduf2019 sorry I'm not able to share the data processing code as it belongs to the company. Did I describe my problem clearly?

**Comment by Saduf2019** [comment]
@shangh1 
In this case, could you please provide us with dummy data with similar shape and size to reproduce the issue reported here.

**Comment by shangh1** [comment]
@Saduf2019 Thanks. I simplify the code, and put it here. https://github.com/shangh1/model_tf2/tree/master

**Comment by shangh1** [comment]
@Saduf2019 @gowthamkpr hello, is there any update on this? Thanks.

**Comment by DachuanZhao** [comment]
> @Saduf2019 @gowthamkpr hello, is there any update on this? Thanks.

Have you read the new document [https://www.tensorflow.org/tutorials/structured_data/preprocessing_layers](https://www.tensorflow.org/tutorials/structured_data/preprocessing_layers) ? It solves my problem .

**Comment by gowthamkpr** [comment]
@shangh1 Please take a look at the above comment.Thanks!

**Comment by DachuanZhao** [comment]
> @shangh1 Please take a look at the above comment.Thanks!

Hi , I want to know whether it will slow down performance  or not in inference status if a model contains a preprocessing layer s  , especially when the model which contains a preprocessing layer served in tf-serving . Is there any document for that ?
